File: app/src/main/java/com/hover/stax/utils/UIHelper.java
package com.hover.stax.utils;

import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.material.snackbar.Snackbar;
import com.hover.stax.R;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import org.jetbrains.annotations.NotNull;

public class UIHelper {
    private static final String TAG = "UIHelper";
    private static final int INITIAL_ITEMS_FETCH = 30;

    public static void flashMessage(Context context, @Nullable View view, String message) {
        if (view == null) flashMessage(context, message);
        else showSnack(view, message);
    }

    private static void showSnack(View view, String message) {
        Snackbar s = Snackbar.make(view, message, Snackbar.LENGTH_LONG);
        s.setAnchorView(view);
        s.show();
    }

    public static void flashMessage(@NotNull Context context, String message) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
    }

    public static void flashMessage(@NotNull Context context, int messageRes) {
        Toast.makeText(context, context.getString(messageRes), Toast.LENGTH_SHORT).show();
    }

    public static LinearLayoutManager setMainLinearManagers(Context context) {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(context);
        linearLayoutManager.setInitialPrefetchItemCount(INITIAL_ITEMS_FETCH);
        linearLayoutManager.setSmoothScrollbarEnabled(true);
        return linearLayoutManager;
    }

    public static int getColor(String hex, boolean isBackground, Context c) {
        try {
            return Color.parseColor(hex);
        } catch (IllegalArgumentException e) {
            return ContextCompat.getColor(c, isBackground ? R.color.offWhite : R.color.brightBlue);
        }
    }

    public static void changeDrawableColor(TextView tv, int color, Context c) {
        for (Drawable d : tv.getCompoundDrawables()) {
            if (d != null)
                d.setColorFilter(new PorterDuffColorFilter(c.getResources().getColor(color), PorterDuff.Mode.SRC_IN));
        }
    }

    static public void setColoredDrawable(ImageButton imageButton, int drawable, int color) {
        Drawable unwrappedDrawable = AppCompatResources.getDrawable(imageButton.getContext(), drawable);
        assert unwrappedDrawable != null;
        Drawable wrappedDrawable = DrawableCompat.wrap(unwrappedDrawable);
        DrawableCompat.setTint(wrappedDrawable, color);
        imageButton.setImageDrawable(wrappedDrawable);
    }

    public static void fixListViewHeight(ListView listView) {
        ListAdapter listAdapter = listView.getAdapter();
        if (listAdapter == null) {
            return;
        }

        int totalHeight = 0;
        int desiredWidth = View.MeasureSpec.makeMeasureSpec(listView.getWidth(), View.MeasureSpec.AT_MOST);
        for (int i = 0; i < listAdapter.getCount(); i++) {
            View listItem = listAdapter.getView(i, null, listView);
            listItem.measure(desiredWidth, View.MeasureSpec.UNSPECIFIED);
            totalHeight += listItem.getMeasuredHeight();
            Log.e(TAG, "item height " + listItem.getHeight());
        }

        ViewGroup.LayoutParams params = listView.getLayoutParams();
        params.height = totalHeight + (listView.getDividerHeight() * (listAdapter.getCount() - 1));
        listView.setLayoutParams(params);
        listView.requestLayout();
    }

    public static int dpToPx(int dp) {
        return (int) (dp * Resources.getSystem().getDisplayMetrics().density);
    }

    public static void loadPicasso(String url, int size, Target target) {
        Picasso.get()
                .load(url)
                .config(Bitmap.Config.RGB_565)
                .resize(size, size).into(target);
    }

    public static void loadPicasso(int resId, int size, Target target) {
        Picasso.get()
                .load(resId)
                .config(Bitmap.Config.RGB_565)
                .resize(size, size).into(target);
    }

    public static void setFullscreenView(Activity activity) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            activity.getWindow().setDecorFitsSystemWindows(false);
        } else {
            activity.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_FULLSCREEN);
        }
    }
}

File: app/src/main/java/com/hover/stax/utils/UIHelper.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.utils

import android.app.Activity
import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Build
import android.view.View
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.Toast
import androidx.annotation.DrawableRes
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.request.target.CustomTarget
import com.google.android.material.snackbar.Snackbar
import com.hover.stax.R
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.FlowCollector
import kotlinx.coroutines.launch

object UIHelper {

    private const val INITIAL_ITEMS_FETCH = 30

    fun showAndReportSnackBar(context: Context, view: View?, message: String) {
        if (view == null) flashAndReportMessage(context, message) else showSnack(view, message)
    }

    private fun showSnack(view: View, message: String?) {
        val s = Snackbar.make(view, message!!, Snackbar.LENGTH_LONG)
        s.anchorView = view
        s.show()
        com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(message, view.context)
    }

    fun flashAndReportMessage(context: Context, messageRes: Int) {
        flashAndReportMessage(context, context.getString(messageRes))
    }

    fun flashAndReportMessage(context: Context, message: String) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
        com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(message, context)
    }

    fun flashAndReportError(context: Context, messageRes: Int) {
        val message = context.getString(messageRes)
        flashAndReportError(context, message)
    }

    fun flashAndReportError(context: Context, message: String) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
        com.hover.stax.core.AnalyticsUtil.logAnalyticsEvent(message, context)
        com.hover.stax.core.AnalyticsUtil.logErrorAndReportToFirebase(context.getString(R.string.toast_err_tag), message, null)
    }

    fun setMainLinearManagers(context: Context?): LinearLayoutManager {
        val linearLayoutManager = LinearLayoutManager(context)
        linearLayoutManager.initialPrefetchItemCount = INITIAL_ITEMS_FETCH
        linearLayoutManager.isSmoothScrollbarEnabled = true
        return linearLayoutManager
    }

    fun setFullscreenView(activity: Activity) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            activity.window.setDecorFitsSystemWindows(false)
        } else {
            activity.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN
        }
    }

    fun ImageView.loadImage(fragment: Fragment, url: String) = GlideApp.with(fragment)
        .load(url)
        .placeholder(R.drawable.icon_bg_circle)
        .circleCrop()
        .override(80)
        .into(this)

    fun ImageView.loadImage(context: Context, url: String) = GlideApp.with(context)
        .load(url)
        .placeholder(R.drawable.icon_bg_circle)
        .circleCrop()
        .override(80)
        .into(this)

    fun ImageView.loadImage(context: Context, @DrawableRes iconId: Int) = GlideApp.with(context)
        .load(iconId)
        .override(100)
        .into(this)

    fun ImageButton.loadImage(context: Context, url: String) = GlideApp.with(context)
        .load(url)
        .placeholder(R.drawable.icon_bg_circle)
        .circleCrop()
        .into(this)

    fun loadImage(context: Context, url: String?, target: CustomTarget<Drawable>) = GlideApp.with(context)
        .load(url)
        .placeholder(R.drawable.icon_bg_circle)
        .circleCrop()
        .override(context.resources.getDimensionPixelSize(R.dimen.logoDiam))
        .into(target)
}

fun <T> Fragment.collectLifecycleFlow(flow: Flow<T>, collector: FlowCollector<T>) {
    viewLifecycleOwner.lifecycleScope.launch {
        viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
            flow.collect(collector)
        }
    }
}